### DISCUSSION TITLE
Installation issue on MacOS

### DISCUSSION THREAD

**Discussion by crbankhurst** [thegetty/quire#102] [Get Help] [answered]
I'm running into a couple of issues with my Quire installation. I am new to using Terminal, so apologies in advance if the solution is painfully obvious! 

Installation of Node.js, Pandoc, Kindle Previewer, and PrinceXML all seemed to go as expected. When I went to install Quire I received the following:

npm WARN deprecated axios@0.19.2: Critical security vulnerability fixed in v0.21.1. For more information, see https://github.com/axios/axios/pull/3410
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules/@thegetty/quire-cli
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules/@thegetty
npm ERR! code EACCES
npm ERR! syscall access
npm ERR! path /usr/local/lib/node_modules/@thegetty/quire-cli
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, access '/usr/local/lib/node_modules/@thegetty/quire-cli'
npm ERR!  [Error: EACCES: permission denied, access '/usr/local/lib/node_modules/@thegetty/quire-cli'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'access',
npm ERR!   path: '/usr/local/lib/node_modules/@thegetty/quire-cli'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.
npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/cbankhurst/.npm/_logs/2021-04-25T17_47_00_129Z-debug.log

Following this, when I type "Quire --help" I do get a list of commands, and I am able to see and work on my-project in Atom. However, I am unable to preview using http://localhost:1313/. Despite having tried various fixes, my browser refuses to connect. I'm unsure whether this is related to the above error or a separate issue entirely. 

Thanks,
Catherine

**Comment by daniel-keller** [comment]
Hi Catherine,
In the error text that you shared I see this line
 `npm ERR! Error: EACCES: permission denied, access '/usr/local/lib/node_modules/@thegetty/quire-cli'`
This means you are not running the command as a user with permission to write to this file location. To correct this you will need to run the command as your computer's administrator account (On Windows this is called the "administrator" user, on MacOS/Linux/Unix this is called the "root" user). 

You can do this by running the same command but prepend the word `sudo`. If you are familiar with Windows, this is the same as right clicking on a program and selecting "Run as Administrator".
```
sudo npm install --global @thegetty/quire-cli
```
You will be prompted to give your password. If your user does have administrator access to the computer quire should install properly. If your user does NOT have administrator access you will need to contact your IT team to get access.

**Reply by crbankhurst** [reply]
Great, thanks so much for walking me through it. That definitely solved that problem, however, at the moment I'm still unable to preview using http://localhost:1313/. I will go back through the steps again now that my Quire install worked properly, and will let you know if I'm still unable to connect.

**Reply by Erin-Cecele** [reply]
Thank you, @daniel-keller. This is a great tip and we have added it to the installation section of our documentation! @crbankhurst Are you still having difficulties getting the preview to work?

**Reply by crbankhurst** [reply]
Hello! Yes, I've just now been having another go at this. I decided to try going back and reinstalling everything to see if I had made a mistake somewhere along the way. All the installs went fine, but I am still getting this error text after typing "quire preview" into Terminal:

Launching preview server
ℹ Navigate to http://localhost:1313 to see your changes.
ℹ Press Control+C or type "quire stop" to stop the preview.
events.js:292
      throw er; // Unhandled 'error' event
      ^

Error: spawn /usr/local/lib/node_modules/@thegetty/quire-cli/node_modules/hugo-bin/vendor/hugo ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:269:19)
    at onErrorNT (internal/child_process.js:465:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
Emitted 'error' event on ChildProcess instance at:
    at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
    at onErrorNT (internal/child_process.js:465:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawn /usr/local/lib/node_modules/@thegetty/quire-cli/node_modules/hugo-bin/vendor/hugo',
  path: '/usr/local/lib/node_modules/@thegetty/quire-cli/node_modules/hugo-bin/vendor/hugo',
  spawnargs: [
    'server',
    '--config=config.yml,config/environments/dev.yml',
    '--watch'
  ]
}
10% building 0/0 modules 0 active
webpack is watching the files…

10% building 0/1 modules 1 active ...ct/themes/default/source/js/application.jsBrowserslist: caniuse-lite is outdated. Please run next command `npm update`
events.js:292
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at afterWriteDispatched (internal/stream_base_commons.js:156:25)
    at writeGeneric (internal/stream_base_commons.js:147:3)
    at Socket._writeGeneric (net.js:785:11)
    at Socket._write (net.js:797:8)
    at writeOrBuffer (internal/streams/writable.js:358:12)
    at Socket.Writable.write (internal/streams/writable.js:303:10)
    at Watching.compilerCallback [as handler] (/Users/cbankhurst/my-project/themes/default/node_modules/webpack-cli/bin/cli.js:335:30)
    at /Users/cbankhurst/my-project/themes/default/node_modules/webpack/lib/Watching.js:99:9
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/cbankhurst/my-project/themes/default/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)
    at AsyncSeriesHook.lazyCompileHook (/Users/cbankhurst/my-project/themes/default/node_modules/tapable/lib/Hook.js:154:20)
    at Watching._done (/Users/cbankhurst/my-project/themes/default/node_modules/webpack/lib/Watching.js:98:28)
    at /Users/cbankhurst/my-project/themes/default/node_modules/webpack/lib/Watching.js:73:19
    at Compiler.emitRecords (/Users/cbankhurst/my-project/themes/default/node_modules/webpack/lib/Compiler.js:499:39)
    at /Users/cbankhurst/my-project/themes/default/node_modules/webpack/lib/Watching.js:54:20
    at /Users/cbankhurst/my-project/themes/default/node_modules/webpack/lib/Compiler.js:485:14
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/cbankhurst/my-project/themes/default/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:9:1)
    at AsyncSeriesHook.lazyCompileHook (/Users/cbankhurst/my-project/themes/default/node_modules/tapable/lib/Hook.js:154:20)
    at /Users/cbankhurst/my-project/themes/default/node_modules/webpack/lib/Compiler.js:482:27
    at /Users/cbankhurst/my-project/themes/default/node_modules/neo-async/async.js:2818:7
    at done (/Users/cbankhurst/my-project/themes/default/node_modules/neo-async/async.js:3522:9)
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/cbankhurst/my-project/themes/default/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at AsyncSeriesHook.lazyCompileHook (/Users/cbankhurst/my-project/themes/default/node_modules/tapable/lib/Hook.js:154:20)
    at /Users/cbankhurst/my-project/themes/default/node_modules/webpack/lib/Compiler.js:464:33
    at /Users/cbankhurst/my-project/themes/default/node_modules/graceful-fs/graceful-fs.js:57:14
    at FSReqCallback.oncomplete (fs.js:171:23)
Emitted 'error' event on Socket instance at:
    at emitErrorNT (internal/streams/destroy.js:106:8)
    at emitErrorCloseNT (internal/streams/destroy.js:74:3)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: -32,
  code: 'EPIPE',
  syscall: 'write'
}

Following this, I can't get localhost to connect. Any thoughts on possible fixes would be much appreciated.

Thanks!
Catherine

**Reply by daniel-keller** [reply]
Hmm, maybe its a version incompatibility issue between quire and nodejs? I've run into trouble if I run older versions of quire on newer versions of node. If you run `node --version` and `quire --version` in the command line what version numbers do you see?

EDIT: I see now the error says `Error: write EPIPE` which sounds like another permissions issue. If you run the quire preview command with sudo does it work?
```
sudo quire preview
```

**Reply by crbankhurst** [reply]
node --version is v14.16.1 and quire --version is v0.19.1

I wondered if it was another permissions issue, but it doesn't work with sudo either. The same "Error: write EPIPE" text occurs.

**Reply by geealbers** [reply]
@crbankhurst I think the first line of the error messaging, `Error: spawn /usr/local/lib/node_modules/@thegetty/quire-cli/node_modules/hugo-bin/vendor/hugo ENOENT` is an indication that Hugo (part of the core Quire installation) did not actually install correctly while other parts of the Quire installation did work.

1. Try running `quire site` and let us know if you see the same error referenced above.
2. Try going back and running `sudo npm install --global @thegetty/quire-cli` one more time, and watch at the end of the process for mentions of Hugo or hugo-bin. There should also be a message that says Hugo was successfully installed.
3. If you do see that successfully installed message, go back to your project file and see if you can now run `quire preview`. If you don't please share the Terminal output and we can try to do some more troubleshooting.

If none of this works, I think we'll want to go in and manually delete the existing files (some might be left over from before you ran with `sudo`) and try again.

**Reply by crbankhurst** [reply]
Thanks very much. I tried steps one and two above and have attached here the terminal messaging. It's quite lengthy! 

[Terminal Saved Output.pdf](https://github.com/thegetty/quire/files/6395105/Terminal.Saved.Output.pdf)

There is some other error messaging, and nothing came up about Hugo at the end. I can go back and manually delete the files as you suggest.

**Reply by daniel-keller** [reply]
I've started with a fresh install to work my way through the setup and here's what I've found: 
* Any time you create a new project you will need to install the default theme's dependencies. Right now `quire new my-new-project` doesn't do this automatically. You have to "cd" into the projects "themes/default" directory and run `npm install`. This will install all the developer build tools (like sass, node-sass, etc) needed to build the site (that's why you see the error "Cannot find module 'node-sass'"). This isn't mentioned in the getting started guide https://quire.getty.edu/documentation/getting-started/. @geealbers Is "quire new" supposed to also install the default theme's npm packages?
* From my testing Quire v0.19.1 doesn't work with the active LTS node version (v14.16.1) but it does with the previous (v12.22.1) which is supported until 2022. Maybe the docs should specify the version of node LTS to use?
* I've got an issue where my system can't locate my XCode build tools. I think this is just an issue on my end though.

**Reply by geealbers** [reply]
@daniel-keller thanks for the extra testing! Running `quire new` is supposed to install the theme dependencies at the same time. We had a problem with this for Windows users, but it should have been fixed and hadn't previously effected Mac users. 

I'm looking into it, but haven't been able to replicate the issue you're seeing with node v14.16.1 and Quire v.0.19.1. For me, those two version are still properly creating and previewing projects, and the theme dependencies are being automatically installed as well. (FYI I've got npm 6.14.12, and I'm running on macOS Catalina v10.15.6.) Any ideas?

@crbankhurst all that aside, I'm going to suggest we maybe jump on a call where you could share your screen and we could try to work through it together? @Erin-Cecele or I will message you shortly.

**Reply by geealbers** [reply]
@daniel-keller 

> I've got an issue where my system can't locate my XCode build tools. I think this is just an issue on my end though.

We have another user with an Xcode issue. She saw `gyp: No Xcode or CLT version detected!` in her Terminal output. I think a fix is suggested at https://medium.com/flawless-app-stories/gyp-no-xcode-or-clt-version-detected-macos-catalina-anansewaa-38b536389e8d if that's of use to you

**Comment by Erin-Cecele** [comment, marked as answer]
Hi @crbankhurst, we have a few more options for you to try. 

Before attempting **_each_** of these options, click on your computer background and press Shift-Command-G. Navigate to `/usr/local/lib` and delete the `node_modules` folder, as we did on our call. Then reinstall the LTS version of nodejs: https://nodejs.org/en/.  

First, we'll try updating the Disk Access settings on your Mac to include Terminal:

1. Navigate to your System Preferences.
2. Click on Security & Privacy. 
3. Choose the Privacy tab at the top right. 
4. Click the lock to make changes.
5. Scroll down the list on the lefthand side until you reach "Full Disk Access." Click on this. 
6. Then on the righthand side, make sure that Terminal is checked. 
7. When this change has been made make sure to click the lock again before closing. 
8. Now try to install again. (Do not use the `sudo npm install` command. Instead use the command given in the Quire install directions: `npm install --global @thegetty/quire-cli`).
9. Run `quire --version` again to make sure the install was successful. 
10. Create a new project and try to preview.

If that doesn't work, we'll try running a different `sudo` command to update permissions for the global node module directory:

1. In terminal, use the `cd` command to get back to your computer's home directly. Then enter the following command: 
```
sudo chown -R $(whoami) $(npm config get prefix)/{lib/node_modules,bin,share}
```
2. Now try steps 8-10 above. 

Let us know if either of these troubleshooting tips works. Thank you!

**Reply by Erin-Cecele** [reply]
If running into permissions errors during the MacOs install process, follow the steps below and use the recommended  `sudo` command to update permissions for the global node module directory:

1. Click on your computer background and press Shift-Command-G. Navigate to `/usr/local/lib` and delete the `node_modules` folder. Then reinstall the LTS version of nodejs: https://nodejs.org/en/.  
2. In Terminal, use the `cd` command to get back to your computer's home directly. Then enter the following command: 
```
sudo chown -R $(whoami) $(npm config get prefix)/{lib/node_modules,bin,share}
```
3. Now try to install again with the command `npm install --global @thegetty/quire-cli`. 
4. Run `quire --version` to make sure the install was successful. 
5. Create a new project with `quire new` and run `quire preview` to view locally.

**Comment by crbankhurst** [comment]
Hi Erin, 

The first option resulted in error messaging, however, the second option worked perfectly and I am now able to preview using http://localhost:1313/. I will let you know if I encounter any more issues, but for now all seems to be working well.

Thank you so much for resolving this for me!

Catherine

**Reply by Erin-Cecele** [reply]
Wonderful news. Thank you for the update, Catherine!

**Reply by leschram** [reply]
Hi @Erin-Cecele, I'm having a similar problem to Catherine. I'm getting the following message when I try to preview my project: 

leslieschramer@Leslies-MacBook-Pro-2 kelsey-marbles % quire preview
Launching preview server
ℹ Navigate to http://localhost:1313 to see your changes.
ℹ Press Control+C or type "quire stop" to stop the preview.
10% building 0/0 modules 0 active
webpack is watching the files…

10% building 0/1 modules 1 active ...es/themes/default/source/js/application.jsBrowserslist: caniuse-lite is outdated. Please run next command `npm update`

I followed the steps outlined above but am getting the same result: I can create a new project and alter the YAML files and the changes are reflected on the localhost, but only in HTML, not rendered in a true preview. Help??

**Reply by Erin-Cecele** [reply]
Hi @leschram. I see @geealbers responded to your related inquiry in discussion #144. As Greg mentioned, please run `quire preview --verbose` and attach your Terminal output. Thank you!
